### PR TITLE
Fix out of sync templates files and add a check

### PR DESCRIPTION
**Pull request by TahaTesser**
### Description
- Add a check to verify template code in the Material library is synced with `gen_defaults`
- Sync the changes to pass the new check.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Review by guidezpl — APPROVED** [review]

**Comment by TahaTesser** [comment]
Will add tests to catch these sync issues in the future.

**Comment by TahaTesser** [comment]
@guidezpl 
From Hixie's feedback I added a test to verif when template code is modified in the Material library files they're in synced with gen_defaults. 

Could you please review the test? or assign someone who could be review this type of test?

**Review by guidezpl — APPROVED** [review]

**Comment by guidezpl** [comment]
Feel free to get a second review

File: dev/bots/analyze.dart
Comment (goderbauer) [inline dev/bots/analyze.dart:255]: `verifyMaterialFilesAreUpToDateWithTemplateFiles` is an async method. Is this line missing an `await`?

**Review by goderbauer — COMMENTED** [review]
This will also need a test or a test exception (see bot message).

File: dev/bots/analyze.dart
Comment (TahaTesser) [reply dev/bots/analyze.dart:255]: Good catch!

**Comment by TahaTesser** [comment]
> This will also need a test or a test exception (see bot message).

The bot message was before I added the `analyze.dart` check. This makes sure that template code in the material library files are synced with `gen_defaults` 

Similar to https://github.com/flutter/flutter/pull/111364. I see there is only `analyze.dart`  check.

**Comment by TahaTesser** [comment]
Add a test in `analayze_test.dart`

**Comment by QuncCccccc** [comment]
Thanks for adding this check!! It was on my todo list but never got implemented...

**Review by goderbauer — COMMENTED** [review]
LGTM, I leave the final review to @QuncCccccc.

File: dev/bots/test/analyze-gen-defaults/dev/tools/gen_defaults/bin/gen_defaults.dart
Comment (TahaTesser) [reply dev/bots/test/analyze-gen-defaults/dev/tools/gen_defaults/bin/gen_defaults.dart:6]: Looks like only relative imports are supporting for these test folders.

I see similar case here:
https://github.com/flutter/flutter/blob/f54c3567e919b726d5b354690c23612b0253b6ba/dev/bots/test/codesign_test.dart#L8-L10

Is there a solution for this?

Comment (TahaTesser) [reply dev/bots/test/analyze-gen-defaults/dev/tools/gen_defaults/bin/gen_defaults.dart]: I've moved the files into the same root folder to avoid the relative lib imports. Credit to @bleroux for the idea.

File: dev/tools/gen_defaults/lib/input_decorator_template.dart
Comment (QuncCccccc) [inline dev/tools/gen_defaults/lib/input_decorator_template.dart:49]: Just curious why we have the change here. When it is focused, we still want to see the hover effect, right? So probably hovered state should be put above focused?

**Review by QuncCccccc — COMMENTED** [review]
Thanks for adding this improvement! Just added one comment below and will take another look later today (sorry still need some time to understand the test for analyser).

File: dev/tools/gen_defaults/lib/input_decorator_template.dart
Comment (TahaTesser) [reply dev/tools/gen_defaults/lib/input_decorator_template.dart:49]: This was a change from another PR which wasn't synced with the template class. I synced it here to pass the new check. 

cc: @bleroux

Comment (QuncCccccc) [reply dev/tools/gen_defaults/lib/input_decorator_template.dart:49]: Ah I see. I found the PR change: https://github.com/flutter/flutter/pull/146127
Looks like this is expected, can we add a comment above so this will not cause any confusion. Usually we'll have a order: pressed->hovered->focused. Here is a little more explanation: https://github.com/flutter/flutter/pull/125905 😄

Comment (bleroux) [reply dev/tools/gen_defaults/lib/input_decorator_template.dart:49]: Hey @QuncCccccc!

My PR which introduced this change, https://github.com/flutter/flutter/pull/146127, landed just before Taha implemented this new analyser test. Thanks to this new test Taha spotted that I forgot to update the template. Instead of me filing a PR to update the template we decided that it will be ok to do it in this PR.

About the change itself, the problem with having the hovered state taking precedence over the focused one is especially noticeable on text field: the focused state for a text field increases border width (2dp) and applies bright colors (primary color or error color) while the hover state has the same border than a non focused text field (1dp) and use a color a little darker that a non focused text field. On desktop, it is also very common that a text field is focused and hovered (when users rely mainly on their mouse), so with the previous implementation when the text field was hovered it feels more or less as it lost focus:

https://github.com/flutter/flutter/assets/840911/a882e39e-8013-4854-bf4e-5ecf3cf3de0e

The M3 specification did not specify the exact behavior for this kind of visual properties impacted by hovered/focused states. What they explain in details is that hover is meant as a layer above the component but that is applicable for filled area not for text style and border like this.

I compared with other M3 implementations, especially https://material-web.dev/components/text-field/, and focus state takes precedence over hovered state for text field border (and also for the label color, that's not the case for Flutter text field for the moment).

I plan to file a similar PR for the label text color, I think it will be great to discuss that topic together on this coming PR, there might be cases I missed.

Comment (QuncCccccc) [reply dev/tools/gen_defaults/lib/input_decorator_template.dart:49]: Got it. This totally makes sense to me! Thanks for the explanation! I think adding a comment above would be helpful because other widgets have a reversed order:)

Comment (bleroux) [reply dev/tools/gen_defaults/lib/input_decorator_template.dart:49]: > Got it. This totally makes sense to me! Thanks for the explanation! I think adding a comment above would be helpful because other widgets have a reversed order:)

To avoid blocking Taha's PR, I added the detailed comment on https://github.com/flutter/flutter/pull/146572

File: dev/bots/test/analyze_test.dart
Comment (QuncCccccc) [inline dev/bots/test/analyze_test.dart:302]: Just a question for my education. Looks like the `chip_template.dart` has the same content as the `chip.dart` in `dev/bots/test/dev/bots/test/analyze-gen-defaults`, why here `shouldHaveErrors` is true?

Comment (TahaTesser) [reply dev/bots/test/analyze_test.dart:302]: Test string description is different. I think i didn't make it different enough.

Comment (QuncCccccc) [reply dev/bots/test/analyze_test.dart:302]: Ah chip_template.dart vs chip.dart! Got it!

**Review by QuncCccccc — APPROVED** [review]
LGTM! Thanks again for the fix:)!

File: dev/bots/test/analyze-gen-defaults/dev/tools/gen_defaults/bin/token_logger.dart
Comment (guidezpl) [inline dev/bots/test/analyze-gen-defaults/dev/tools/gen_defaults/bin/token_logger.dart:1]: seems a bit strange to copy this file, is this common?

Comment (TahaTesser) [reply dev/bots/test/analyze-gen-defaults/dev/tools/gen_defaults/bin/token_logger.dart:1]: This is just for the test folder. Generator in the `gen_defaults` cannot update test file in the `dev/bots`

There are similar test folders in the `dev/bots` folder 

https://github.com/flutter/flutter/tree/master/dev/bots/test/analyze-test-input/root/packages/flutter/lib

**Pull request merged**
